nexus/src/app/background/tasks/instance_reincarnation.rs

```diff
@@ -0,0 +1,715 @@
+                        "failed to list instances in need of reincarnation";
+                        "error" => &error,
+                    );
+                    status.query_error = Some(error.to_string());
+                    break;
+                }
+            };
+
+            paginator = p.found_batch(&batch, &|instance| instance.id());
+
+            let found = batch.len();
+            if found == 0 {
+                debug!(
+                    opctx.log,
+                    "no more instances in need of reincarnation";
+                    "total_found" => status.instances_found,
+                );
+                break;
+            }
+
+            let prev_sagas_started = tasks.len();
+            status.instances_found += found;
+
+            let serialized_authn = authn::saga::Serialized::for_opctx(opctx);
+            for db_instance in batch {
+                let instance_id = db_instance.id();
+
+                // Don't reincarnate instances that were restarted too recently.
+                //
+                // Note that we *could* filter out instances that reincarnated
+                // too recently in the `find_reincarnatable_instances` database
+                // query, but I thought it was nice to be able to include them
+                // in this background task's OMDB status.
+                if let Some(last_reincarnation) =
+                    db_instance.runtime().time_last_auto_restarted
+                {
+                    // TODO(eliza): allow overriding the cooldown at the project
+                    // level, once we implement
+                    // https://github.com/oxidecomputer/omicron/issues/1015
+                    if Utc::now().signed_duration_since(last_reincarnation)
+                        < self.default_cooldown
+                    {
+                        status
+                            .instances_cooling_down
+                            .push((instance_id, last_reincarnation));
+                        debug!(
+                            opctx.log,
+                            "instance still needs to take some time to cool \
+                             down before its next reincarnation";
+                            "instance_id" => %instance_id,
+                            "last_reincarnated_at" => %last_reincarnation,
+                        );
+                        continue;
+                    }
+                }
+
+                let prepared_saga = instance_start::SagaInstanceStart::prepare(
+                    &instance_start::Params {
+                        db_instance,
+                        serialized_authn: serialized_authn.clone(),
+                        reason: instance_start::Reason::AutoRestart,
+                    },
+                );
+                match prepared_saga {
+                    Ok(saga) => {
+                        let start_saga = self.sagas.clone();
+                        tasks.spawn(async move {
```

Comment:
Yeah, so we mentioned this offline, but some concurrency bound would be nice here.

Reply:
Added a concurrency limit in https://github.com/oxidecomputer/omicron/pull/6503/commits/cd78516b2341c4117b388164877b16b639042b4c. Note that it depends on merging #6565 first, which --- it turns out --- is actually necessary for _any_ background task that starts sagas to correctly limit their concurrency. Previous implementations were just limiting the concurrency of _starting_ sagas in parallel, which wasn't actually what we care about, really. Whoopsie.